File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-cs/Controllers/SuppliersController.cs
using Microsoft.AspNetCore.Mvc;
using OctoSupply.Api.Models;
using OctoSupply.Api.Repositories;
using OctoSupply.Api.Utils;

namespace OctoSupply.Api.Controllers;

[ApiController]
[Route("api/suppliers")]
public sealed class SuppliersController(SuppliersRepository repository) : ControllerBase
{
    private readonly SuppliersRepository _repository = repository;

    [HttpGet]
    public async Task<ActionResult<IReadOnlyList<Supplier>>> GetAll() =>
        Ok(await _repository.FindAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<Supplier>> GetById(int id)
    {
        var item = await _repository.FindByIdAsync(id);
        return item is null ? throw new NotFoundException("Supplier", id) : Ok(item);
    }

    [HttpGet("{id:int}/status")]
    public async Task<ActionResult<object>> GetStatus(int id)
    {
        var supplier = await _repository.FindByIdAsync(id);
        if (supplier is null)
        {
            throw new NotFoundException("Supplier", id);
        }

        // Mirrors TypeScript behavior where status currently always resolves to APPROVED.
        return Ok(new { status = "APPROVED" });
    }

    [HttpPost]
    public async Task<ActionResult<Supplier>> Create([FromBody] CreateSupplierRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Name))
        {
            throw new ValidationException("name is required");
        }

        var created = await _repository.CreateAsync(request);
        return CreatedAtAction(nameof(GetById), new { id = created.SupplierId }, created);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<Supplier>> Update(int id, [FromBody] UpdateSupplierRequest request) =>
        Ok(await _repository.UpdateAsync(id, request));

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        await _repository.DeleteAsync(id);
        return NoContent();
    }
}

File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-java/src/main/java/com/octocat/supply/controller/SuppliersController.java
package com.octocat.supply.controller;

import com.octocat.supply.exception.NotFoundException;
import com.octocat.supply.model.SupplierModel.CreateSupplierRequest;
import com.octocat.supply.model.SupplierModel.Supplier;
import com.octocat.supply.model.SupplierModel.UpdateSupplierRequest;
import com.octocat.supply.repository.SuppliersRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/suppliers")
public class SuppliersController {
    private final SuppliersRepository repository;

    public SuppliersController(SuppliersRepository repository) {
        this.repository = repository;
    }

    @GetMapping
    public List<Supplier> getAll() {
        return repository.findAll();
    }

    @GetMapping("/{id}")
    public Supplier getById(@PathVariable int id) {
        return repository.findById(id).orElseThrow(() -> new NotFoundException("Supplier not found"));
    }

    @GetMapping("/{id}/status")
    public Map<String, String> getStatus(@PathVariable int id) {
        repository.findById(id).orElseThrow(() -> new NotFoundException("Supplier not found"));
        return Map.of("status", "APPROVED");
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Supplier create(@RequestBody CreateSupplierRequest request) {
        ControllerValidation.requireText(request.name(), "name");
        return repository.create(request);
    }

    @PutMapping("/{id}")
    public Supplier update(@PathVariable int id, @RequestBody UpdateSupplierRequest request) {
        if (!repository.findById(id).isPresent()) {
            throw new NotFoundException("Supplier not found");
        }
        repository.update(id, request);
        return repository.findById(id).orElseThrow(() -> new NotFoundException("Supplier not found"));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable int id) {
        if (!repository.delete(id)) {
            throw new NotFoundException("Supplier not found");
        }
    }
}
